Use dts to simplify and refactor code of STM32 ADC driver

gautierg-st opened the pull request:
This PR is an attempt at simplifying the STM32 ADC driver and reduce code clutter.
The idea is to move as much info as possible to the dts to reduce the need of specific code in the driver.
This PR moves the ADC resolution to the dts for all series.

erwango commented:
@heinwessels @xyzzy42 Can you have a look ?

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -1431,14 +1281,12 @@ static const struct adc_stm32_cfg adc_stm32_cfg_##index = {		\
 	.has_temp_channel = DT_INST_PROP(index, has_temp_channel),	\
 	.has_vref_channel = DT_INST_PROP(index, has_vref_channel),	\
 	.has_vbat_channel = DT_INST_PROP(index, has_vbat_channel),	\
```
I don't think we need these prop anymore. Do we ?

gautierg-st replied:
I suppose we could remove them and make temp_channel & others default to -1 to prevent unwanted behavior

gautierg-st replied:
Done

xyzzy42 commented on dts/bindings/adc/st,stm32-adc.yaml:
```diff
@@ -44,5 +44,26 @@ properties:
     type: boolean
     description: Indicates if the ADC has a dedicated internal vbat monitoring channel.
 
+  temp-channel:
```
The new properties here seem to make the previous has-xxx properties redundant.  It's also possible to create an invalid dts that doesn't produce a warning.  E.g.

```
/* no has-vref-channel */
vref-channel = <10>;
```
Or:
```
has-vref-channel;
has-vbat-channel;
/* no vref-channel nor vbat-channel.  So they are both on channel 0? */
```

If the backward compatibility rules don't make it too onerous, you could get rid of the old has-xx properties.  In the code, use:
```
.vref_channel = DT_INST_PROP_OR(index, vref_channel, 0xff),
```
Where 0xff is a code to indicate no vref channel.

You could also keep boolean fields in the config struct, but they seem unnecessary, like this:
```
.has_vref_channel = DT_INST_NODE_HAS_PROP(index, vref_channel),
.vref_channel = DT_INST_PROP_OR(index, vref_channel, 0),
```

gautierg-st replied:
You're right, it is cleaner this way. Done

xyzzy42 commented on drivers/adc/adc_stm32.c:
```diff
@@ -263,9 +239,11 @@ struct adc_stm32_cfg {
+	int8_t vref_channel;
+	int8_t vbat_channel;
+	int8_t res_table_size;
+	const int32_t *res_table;
```
It looks like the values were small enough to use a `uint16_t` here.

gautierg-st replied:
Fixed

erwango commented on dts/arm/st/wb/stm32wb55.dtsi:
```diff
@@ -5,3 +5,11 @@
+
+/ {
+	soc {
+		adc1: adc@50040000 {
```
Isn't it common to all WB SoC ?

gautierg-st replied:
No, for example, it is on channel 12 for STM32WB15xx

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -609,64 +558,26 @@ static void adc_stm32_teardown_channels(const struct device *dev, uint8_t channe
+	if (config->temp_channel == channel_id) {
 		adc_stm32_disable(adc);
 		adc_stm32_unset_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
+		k_usleep(LL_ADC_DELAY_TEMPSENSOR_STAB_US);
```
This addition is not described on commit message.
Can you split this to another commit with proper justification ?

gautierg-st replied:
This was a bad copy paste. Fixed

erwango commented on include/zephyr/dt-bindings/adc/stm32h7_adc.h:
```diff
@@ -0,0 +1,27 @@
+#define STM32_ADC_RES_10		STM32_ADC_RES(10, 3, 2)
+
+/* 8-bit resolution has a different value for various MCUs */
+#define STM32_ADC_RES_8_H72x_3x_Ax_Bx	STM32_ADC_RES(8, 7, 2)
+#define STM32_ADC_RES_8_H74x_5x		STM32_ADC_RES(8, 4, 2)
```
I would prefer STM32H72x_ADC_RES_8 and STM32H74x_ADC_RES_8

gautierg-st replied:
Done

xyzzy42 commented on dts/arm/st/h7/stm32h723.dtsi:
```diff
@@ -40,6 +40,32 @@
+		};
+
+		adc3: adc@58026000 {
+			st,resolutions = <STM32_ADC_RES_12 STM32_ADC_RES_10
+					  STM32H72x_ADC_RES_8
+					  STM32H72x_ADC3_RES_6>;
```
Isn't the shift value wrong for the resolutions other than 6 bits?  It looks like ADC3 should use a shift of 3 for every resolution supported.

> 
> Bits 4:3 RES[1:0]: Data resolution
> These bits are written by software to select the resolution of the conversion.
> 00: 12-bit
> 01: 10-bit
> 10: 8-bit
> 11: 6-bit
>

xyzzy42 commented:
It occurs to me that the shift value for the resolutions should probably be the same for every resolution supported by an ADC.  At least, this appear to be the case for every ADC header's definitions except one, and I think that is a bug.

Given that, it could be moved to its own property and not be an array.  `st,resolution-shift = <3>` for example.

With that, the resolution field fits in 8-bits instead of 16.

It then also looks like the register field for each resolution in sequentially assigned, or nearly so.  I.e., the first resolution is always 0, the next 1, and so on.  They are the same value as the position in the resolutions array.  So it seems like it should be possible to not even have the resolution coding/decoding and just put the straight value in there.

Example for STM32H72x, the complex one, most are much simpler:

```
&adc1_2  {
	// This one has a gap for "non-legacy mode" resolutions
	st,resolutions = <16 14 12 10 0 0 0 8>;
	st,resolution-shift = <2>;
};
adc3: adc@58026000 {
	st,resolutions = <12 10 8 6>;
	st,resolution-shift = <3>;
};
```

When looping through the resolution array, one just needs to skip the `0` values, in I think just one ADC, to account for gaps in the assignments.

gautierg-st commented:
As said above, the only case where the offset is not the same for a specific ADC is the ADC3 on STM32H72x and the reason is that we use the value in the dts to recreate some `#define` found in the LL header.

In this case:
```
adc3: adc@58026000 {
	st,resolutions = <12 10 8 6>;
	st,resolution-shift = <3>;
};
```
For the the function `LL_ADC_SetResolution` we need to pass 
- (0b010 << 2) for 12-bit resolution
- (0b011 << 2) for 10-bit resolution
- (0b111 << 2) for 8-bit resolution
- (0b11 << 3) for 6-bit resolution (that could potentially be (0b110 << 2)

Maybe we could set `st,resolutions = < 0 0 12 10 0 0 6 8>;` in this case (with a shift to 2), but I'm not sure it is easier or clearer.

xyzzy42 commented:
Is it worth it to use `LL_ADC_SetResolution`?  IMHO, it's quite confusing that the dts appears to encode the values used in the datasheet, but it doesn't really because the HAL does modifications to them in some cases.

I think the two HALs, Zephyr dts and ST, is just making it more complex.

The dts here can do a better job of encoding the differences between ADC versions.  It does the job `LL_ADC_SetResolution` is trying to do, but does it better, because it works across families.  It looks like it could do everything done in the HAL very easily.  In fact, the Zephyr code and dts would be simpler, since it wouldn't need to generate unusual values for the resolution code.

If we look at this ADC, the values we need to program it are:
`{0 << 3, 1 << 3, 2 << 3, 3 << 3}`
That's pretty simple.  It's easy and efficient to get from the dts.  And it matches the datasheet.

But with the ST HAL, we need to instead encode:
`{2 << 2, 3 << 2, 7 << 2, 3 << 3}`
And then at runtime some bit manipulation code in the HAL turns this sequence into the previous simpler one.

It's also worth considering that the dts is not really describing the hardware here.  Rather, it's describing the HAL macro values!  So if the value of `LL_ADC_RESOLUTION_xxB` changes in the ST HAL, then the Zephyr dts files become wrong.

I thought it was describing the hardware.  It almost is, except in some cases.  I wonder how many other reviewers realized that the dts was encoding macro values from the HAL and not the ADCs' registers?

gautierg-st commented:
I agree that we should keep the dts to describe hardware values. But if we don't use the SetResolution function and write the register directly then we need to describe it also.
For STM32H72x, it would be something like this:
```
adc1: adc@40022000 {
	st,resolution-reg = < 0xC 2 3 > /* regiser address offset, resolution offset, resolution width */
	st,resolutions = < 16 14 12 10 0 0 0 8 >; /* List of all supported resolutions */
};
```
That way we describe the hardware exactly, and we can get rid of the SetResolution function.
Please let me know what you think.
Also, I will split this PR to keep the work on resolution and on internal channels separated.

xyzzy42 commented:
> I agree that we should keep the dts to describe hardware values. But if we don't use the SetResolution function and write the register directly then we need to describe it also. For STM32H72x, it would be something like this:
> 
> ```
> adc1: adc@40022000 {
> 	st,resolution-reg = < 0xC 2 3 > /* regiser address offset, resolution offset, resolution width */
> 	st,resolutions = < 16 14 12 10 0 0 0 8 >; /* List of all supported resolutions */
> };
> ```

Generally dts files work better if each property does one thing.  So reg address, offset of field, width of field would all be individual properties.

IMHO, not that I'm the authority here, but I like this better than using the HAL.  
* The code in Zephyr is smaller, simpler, and more efficient.
* The dts binding is simpler and reflects the chip, rather than a HAL that is subject to change.

Another way to do it, would be to write an inline function that sort of like SetResolution, which only sets the field.
```
void stm32_adc_set_resolution(ADC_TypeDef *ADCx, uint32_t resolution)
{
        MODIFY_REG(ADCx->CFGR1, ADC_CFGR1_RES, Resolution);
}
```
But you need to change the names of the struct field and bitmask for each family.  Don't do any of the other stuff to remap values.  It's sort of a combo of dts + family specific inline functions to get abstraction.  I don't know if this is actually better.  Maybe if there already was a set of headers in Zephyr to abstract the HAL differences?  A hardware abstraction layer abstraction layer, or HALAL.

The HAL is just really annoying!
There is a define for the register field offset, but it's in a header with typedefs so it can't be used from a dts.
There is no define for the register field width, just a bitmask.
There is no define for the register address.  Just a struct.
The struct name, field, and macro names are different for each chip.
There's no function with a common API to access a common field, except...
The SetResolution() function doesn't just access the field, it also remaps the values, but..
The remapping doesn't abstract the differences between the families, rather ironically, the abstraction layer is more complex than the hardware it abstracts.

gautierg-st commented:
Rebased now that #54038 is merged. No modification from previous code. Waiting for more feedback about the best way to go.

erwango commented:
> But you need to change the names of the struct field and bitmask for each family. Don't do any of the other stuff to remap values. It's sort of a combo of dts + family specific inline functions to get abstraction. I don't know if this is actually better. Maybe if there already was a set of headers in Zephyr to abstract the HAL differences? A hardware abstraction layer abstraction layer, or HALAL.

I'd prefer we avoid starting maintaining new type of HAL.
The goal for using dts is to simplify the code if this can be done with existing LL function that's good, if some zephyr dedicated functions should be added, I'm fine with it but it should be future proof to the maximum extent.. we should not need to add new functions/code when new series (with potentially new registers names will be added. At least LL API are generally good on this part).

erwango commented:
> The goal for using dts is to simplify the code if this can be done with existing LL function that's good, if some zephyr dedicated functions should be added, I'm fine with it but it should be future proof to the maximum extent.

Let's go with the full device tree version

gautierg-st commented on drivers/adc/adc_stm32.c:
```diff
@@ -997,17 +971,17 @@ static int start_read(const struct device *dev,
 #if defined(CONFIG_SOC_SERIES_STM32G0X) || \
 	defined(CONFIG_SOC_SERIES_STM32WLX)
-	if (LL_ADC_GetResolution(adc) != resolution) {
+	if (get_resolution(dev) != resolution) {
 		/*
 		 * Writing ADC_CFGR1 register while ADEN bit is set
 		 * resets RES[1:0] bitfield. We need to disable and enable adc.
 		 */
 		adc_stm32_disable(adc);
-		LL_ADC_SetResolution(adc, resolution);
+		set_resolution(dev, resolution);
 	}
 #elif !defined(CONFIG_SOC_SERIES_STM32F1X) && \
 	!defined(STM32F3X_ADC_V2_5)
-	LL_ADC_SetResolution(adc, resolution);
+	set_resolution(dev, resolution);
 #endif
```
I haven't touched the logic here, but I hesitate to remove all #if, add a simple if to check whether config->res_reg is equal to 0xFF (to filter out STM32F1 and F334), and to apply the same logic of read, disable, set for all other series.
Any thoughts?

FRASTM commented:
> > The goal for using dts is to simplify the code if this can be done with existing LL function that's good, if some zephyr dedicated functions should be added, I'm fine with it but it should be future proof to the maximum extent.
> 
> Let's go with the full device tree version

Is a format like include/zephyr/dt-bindings/clock/stm32xx_clock.h  possible for the series : 
```
/** @brief ADC_CFGR register offset */
#define ADC_CFGR		0x0C

```
```
plus a macro to retrieve the resolution : 
STM32_RES(val, 3, 3, ADC_CFGR)

````

erwango commented:
> > > The goal for using dts is to simplify the code if this can be done with existing LL function that's good, if some zephyr dedicated functions should be added, I'm fine with it but it should be future proof to the maximum extent.
> > 
> > 
> > Let's go with the full device tree version
> 
> Is a format like include/zephyr/dt-bindings/clock/stm32xx_clock.h possible for the series . Example for stm32l1xx :
> 
> ```
> /** @brief ADC_CFGR register offset */
> #define ADC_CFGR		0x0C
> ```
> 
> plus a macro to retrieve the resolution : `STM32_RES(val, 3, 3, ADC_CFGR)`
> 
> or for stm32l1xx : `#define ADC_CFGR 0x04`
> 
> `STM32_RES(val, 3, 24, ADC_CFGR)`

This was the initial proposal. Current solution, is the result of discussions starting from https://github.com/zephyrproject-rtos/zephyr/pull/53868#issuecomment-1400456521

gautierg-st commented:
Hi @xyzzy42, I would be happy to have your feedback on this new version

gautierg-st committed:
include: zephyr: dt-bindings: adc: add macros for stm32 adc resolutions

Define some macros that will help to define the STM32 ADC resolutions in
the dtsi files. The end goal is to simplify the driver by removing code
specific to particular series.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
dts: bindings: adc: add new dts bindings for stm32 adc resolution

Add a binding for STM32 ADC to specify the resolutions and all associated
register information (through a STM32_ADC_RES macro).

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
dts: arm: st: add adc resolutions property in all stm32 dtsi

Add new ADC resolution property in all STM32 dtsi files, for all ADC
instances.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: get resolutions from dtsi for stm32 adc

Simplify the STM32 ADC driver code by using the new ADC resolutions
property in dtsi files.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st commented:
New version using macros in dtsi again, but without using LL functions. That still simplifies the driver, but does not overcharge the dtsi properties.

xyzzy42 commented on include/zephyr/dt-bindings/adc/stm32_adc.h:
```diff
@@ -0,0 +1,57 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_STM32_ADC_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_STM32_ADC_H_
+
+#define STM32_ADC_REG_MASK		0xFFU
```
You could use `BIT_MASK(8)` and so on.  This macro is already there for Zephyr ADC bindings in `include/zephyr/dt-bindings/adc/adc.h`.  But maybe this header is annoying to include here and we certainly all know what a bitmask in hex is.

gautierg-st replied:
Done

xyzzy42 commented on include/zephyr/dt-bindings/adc/stm32f4_adc.h:
```diff
@@ -0,0 +1,19 @@
+#define RES_SHIFT	24
+#define RES_MASK	0x03
+
+#define STM32_ADC_RES(resolution, reg_val)	\
+	STM32_ADC(resolution, reg_val, RES_MASK, RES_SHIFT, CFGR_REG)
```
This macro could be moved into `stm32_adc.h` and not duplicated in each chip specific header.  The only difference would be on stm32f1, where `STM32_ADC_RES(12)` would become `STM32_ADC_RES(12, 0x00);`.  Maybe that's not quite as good for f1?  But it's a nice refactor otherwise and only a small change on the f1.

The H7 header would still have the special adc3 version.

gautierg-st replied:
Done

xyzzy42 commented on drivers/adc/adc_stm32.c:
```diff
@@ -828,72 +806,89 @@ static void dma_callback(const struct device *dev, void *user_data,
+
+	uintptr_t addr = (uintptr_t)adc + reg;
+
+	sys_clear_bits(addr, mask << shift);
+	sys_set_bits(addr, value << shift);
```
This will do two register writes.  Probably don't want to do that, as writing to the register has side effects.  There are two significant issues with this.

* The compiler can not optimize out the intermediate write.  Instead of generating efficient code to clear and set the field in the a CPU register, then write to the device, it must first clear the field, write this to the device register, then read it back from the device register, then set the bits, then write it again.
* Setting the field to all zeros in the device register might actually do something on the device!

Every decent API for reading or writing register fields I've used or designed includes a function that will set a field to a new value, without multiple writes to the register.  Example from the Linux regmap API:

```c
 regmap_update_bits(struct regmap *map, unsigned int reg, unsigned int mask, unsigned int val);

regmap_update_bits(adc, reg, mask << shift, value << shift);
```

Functions named `*and_or*` is another common pattern for this.  Since the mask is and-ed and the the value is or-ed.

But the Zephyr functions in `sys_bitops.h` don't seem to have it.

gautierg-st replied:
Agreed, uses MODIFY_REG instead

xyzzy42 commented on drivers/adc/adc_stm32.c:
```diff
@@ -296,6 +272,8 @@ struct adc_stm32_cfg {
 	int8_t temp_channel;
 	int8_t vref_channel;
 	int8_t vbat_channel;
+	int8_t res_table_size;
+	const uint32_t *res_table;
```
This can be packed a little more efficiently, without the extra pointer, by using a flexible array member, which is part of the C99 standard, via:

```c
struct adc_stm32_cfg {
	ADC_TypeDef *base;
	…
	int8_t res_table_size;
	const uint32_t res_table[];
};

static const struct adc_stm32_cfg adc_stm32_cfg_##index = {		\
	…
	.res_table_size = DT_INST_PROP_LEN(index, resolutions),		\
	.res_table = DT_INST_PROP(index, resolutions),			\
};									\

gautierg-st replied:
Done

gautierg-st commented:
Fixed CI, fix conflicts, and adds STM32C0 resolution now that it is merged

GeorgeCGV commented:
Regarding ADC:

https://github.com/zephyrproject-rtos/zephyr/blob/6bbadb6012cc04b9b9eaa14c89571e33c91583a6/drivers/adc/adc_stm32.c#L676-L687

What could be the case that `LL_ADC_Enable` done before `LL_ADC_IsActiveFlag_ADRDY` loop might fail and require an extra `LL_ADC_Enable` check and call?

gautierg-st commented:
> What could be the case that `LL_ADC_Enable` done before `LL_ADC_IsActiveFlag_ADRDY` loop might fail and require an extra `LL_ADC_Enable` check and call?

For some series, if a calibration is on-going, the ADC cannot be enabled, hence the retry. The relevant part of the reference manual is this one:
![image](https://user-images.githubusercontent.com/103205175/233964433-4194dc97-6641-4b03-8760-a71a19204d69.png)
It is present for example for STM32G4 (RM0440), STM32L5 (RM0438) or STM32WB (RM0434), and some others.

I'm not sure it could happen in the current code though since we check and wait the end of the calibration.
Also, some families (like H5) should also execute this bit of code (but currently don't) as they have same on-off procedure.
I agree that this code definitely needs some rework but that is not the object of the current PR.

gautierg-st commented on dts/bindings/adc/st,stm32-adc.yaml:
```diff
@@ -44,5 +44,13 @@ properties:
+    description: |
+      List of the resolutions supported by the ADC instance. They should be
+      values created with STM32_ADC_RES macro or similar. Their order is not
+      important.
```
Done

erwango commented:
@xyzzy42, @GeorgeCGV Would you mind having a new look ?

GeorgeCGV commented on include/zephyr/dt-bindings/adc/stm32h7_adc.h:
```diff
@@ -0,0 +1,27 @@
+ * The offset and the width of the resolution field need to be redefined.
+ */
+#define STM32H72X_ADC3_RES_SHIFT	3
+#define STM32H72X_ADC3_RES_MASK		0x03
```
Why not `BIT_MASK(2)` for consistency?
